### ISSUE TITLE
@byCedric

### ISSUE THREAD

**Issue by HYJ29** [byCedric/office-marathon#7]
@byCedric 
I really thank you for your effort. 👍 
But the problem keeps happening and I couldn't find anything from google for several days.. So I came back again.. 
Would you try 
1. timeInterval: 1000, 
2. android 10 simulator 
3. push start tracking 
4. turn off the simulator to sleep mode 
5. and wait 5 minutes? 

I'm sure you can reproduce it! 

## just background.. 

I set the time as 1000 ms ( because I need to track a running person ) 
And I think this work on ios, not terminating process or app on sleep mode ( I tested about 1hour ) 

but on Android
On Android 10 simulator, this stops on sleep mode. For me, it logs about 72 locations and stops. 
If I get out from sleep mode, It keeps logging.. ( for some reason It goes crazy.. )
The main problem is it stops tracking location on sleep mode.. 

and
for some reason, I think expo SDK40 gives me a permission error like this below 
https://forums.expo.io/t/sdk40-android-unhandled-promise-rejection-error-not-authorized-to-use-background-location-services/46508/4

This makes me really exhausting. I want to know is it about google policy So I should wait for a while until expo team deals with it? or it's just a problem of me?

_Originally posted by @HYJ29 in https://github.com/byCedric/office-marathon/issues/2#issuecomment-753959693_

**Comment by HYJ29** [comment]
@byCedric 
Finally, I found a solution. I want to share it with you and hear your opinion. I want to be convinced that there is no problem with it.

## Problem 
The problem was after some minutes phone enters sleep mode, It stops logging ( on timeInterval 1000ms ) 
I thought it was the termination of tasks, but it was not. 
**If I reactive phone after the logging stops,  The TaskManager gives me an array of locations it tracked, but not logged. 
for example, If the log stops around 10 seconds, it gives 10 location object on locations array.**

So for ideally, it should have no problem on the code and should work right.
But for some reason, if I have too much item to iterate to 'addLocation', 
the number of locations that TaskManager give me increase by one every second
So I mean, if I reactive phone after 40seconds since it stops logging, 
TaskManger give me 40 locations array at first, but 41 locations after once second, 42 locations after 2 second.. so on 
that was the reason why logging goes crazy after reactivating the phone. 


## Solutions
So I fixed it to not iterate, but just dump in the whole locations object like this 
`const newLocations = [...existingLocations, ...locations ] `

I didn't check that the locations object is exactly the locations that the phone tracked, but I think it would be, I'll test it soon. 

I will request PR about this. 
It would be really appreciated for me If you look into it and answer me.

Thank you!

**Comment by Javierko** [comment]
Does this work in production?

I have same problem as you described, when I'm in Expo and developing my app and lock my phone, then location works fine, also it fetched not by minute but for example 5 or 10 minutes or when i unlock phone it send me locations when phone was locked.

But I couldn't make it working on production mode in standalone app (made with `expo build:android`). Like when I lock the phone, it doesn't fetch anything and when I unlock the phone, it sends only one location with the time when I unlocked the phone.

**Comment by vasildimitroffhristoff** [comment]
@byCedric this seems like a valid issue. We are able to reproduce it on a few real devices as well (Android 10). 

@HYJ29 Did you manage to fix this issue?

**Comment by HYJ29** [comment]
@byCedric 
I could fix the original issue with the above workaround.
but I still have an issue. I'm not working on it right now. But want to share with you.
this is my published app to the google play store for an internal test. (ios works well) 
I run it with android galaxy 8, android version 9. 

I think the arrays, android giving to me while on sleep mode are not accurate. And I need to look into it.

as you see below, there are some duplicated points recorded. 
<img src="https://user-images.githubusercontent.com/44938459/114321707-17483400-9b57-11eb-87bc-ca4a231e0f06.png" width="500" />

I also recorded with 'Strava'(a popular app for running tracking ). We can see the difference comparing with it.
<img src="https://user-images.githubusercontent.com/44938459/114321831-e0265280-9b57-11eb-86ab-4a5faa343fe4.png" width="500" />

**Comment by vasildimitroffhristoff** [comment]
@HYJ29 Some nasty fluctuating points appear especially with Android. Have you tried any filtering of the noisy data such as Kalman filter?

**Comment by HYJ29** [comment]
@vasildimitroffhristoff No, I have not. I don't have much knowledge of filtering raw data. Do you think it would be the solution~?

**Comment by vasildimitroffhristoff** [comment]
@HYJ29 we found out that the array with the locations coming from the background task after the device is awaken from sleep are in messed up order. Try sorting them by timestamp. I think it will fix some of the issues with the strange lines appearing on your track.

**Comment by HYJ29** [comment]
@vasildimitroffhristoff  Thank you for letting me know! I'll try that
